src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -34640,6 +34640,519 @@ .maxstack  2
+        {
+            var source = """
+                using System.Collections;
+                using System.Collections.Generic;
+
+                class MyCollection(List<int> list) : ICollection<int>
+                {
+                    public int Count => list.Count;
+
+                    public bool IsReadOnly => false;
+
+                    public void Add(int item) => list.Add(item);
+
+                    public void Clear() => list.Clear();
+
+                    public bool Contains(int item) => list.Contains(item);
+
+                    public void CopyTo(int[] array, int arrayIndex) => list.CopyTo(array, arrayIndex);
+
+                    public bool Remove(int item) => list.Remove(item);
+
+                    public IEnumerator<int> GetEnumerator() => list.GetEnumerator();
+
+                    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+                }
+
+                class C
+                {
+                    static void Main()
+                    {
+                        M(new([1, 2, 3])).Report();
+                    }
+
+                    static List<int> M(MyCollection c) => [..c];
+                }
+                """;
+
+            var verifier = CompileAndVerify([source, s_collectionExtensionsWithSpan], expectedOutput: IncludeExpectedOutput("[1, 2, 3],"), targetFramework: TargetFramework.Net80, verify: Verification.Skipped);
+            verifier.VerifyDiagnostics();
+
+            verifier.VerifyIL("C.M", """
+                {
+                  // Code size       89 (0x59)
+                  .maxstack  3
+                  .locals init (int V_0,
+                                System.Collections.Generic.List<int> V_1,
+                                System.Span<int> V_2,
+                                int V_3,
+                                System.Collections.Generic.IEnumerator<int> V_4,
+                                int V_5)
+                  IL_0000:  ldarg.0
+                  IL_0001:  dup
+                  IL_0002:  callvirt   "int MyCollection.Count.get"
+                  IL_0007:  stloc.0
+                  IL_0008:  ldloc.0
+                  IL_0009:  newobj     "System.Collections.Generic.List<int>..ctor(int)"
+                  IL_000e:  stloc.1
+                  IL_000f:  ldloc.1
+                  IL_0010:  ldloc.0
+                  IL_0011:  call       "void System.Runtime.InteropServices.CollectionsMarshal.SetCount<int>(System.Collections.Generic.List<int>, int)"
+                  IL_0016:  ldloc.1
+                  IL_0017:  call       "System.Span<int> System.Runtime.InteropServices.CollectionsMarshal.AsSpan<int>(System.Collections.Generic.List<int>)"
+                  IL_001c:  stloc.2
+                  IL_001d:  ldc.i4.0
+                  IL_001e:  stloc.3
+                  IL_001f:  callvirt   "System.Collections.Generic.IEnumerator<int> MyCollection.GetEnumerator()"
```

Comment:
I am confused by this codegen. Because it implements `ICollection<int>`, I expected `List<int>.AddRange(IEnumerable<int>)` to be used here.

Reply:
For this case we don't even get to `Add` vs `AddRange` method choice, because we fall into [this Span-based optimization branch](https://github.com/dotnet/roslyn/blob/3328d049728aa83db1af20cf5a1630ad83d946e6/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs#L984).

I suggest to leave it as is because a) it seems out of scope of the precise issue we try to fix (maybe we need to additionally clarify when to use span-based vs `AddRange`-based optimization since the suggestion itself is quite reasonable, but that is another workitem IMO) and b) the next optimization on my list after this one is merged is a [ToList optimization for a single spread](https://github.com/dotnet/roslyn/issues/71217) (no pun intended), so this will turn into a single `ToList` call anyway